src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -672,71 +678,82 @@ const static auto formats = {
+protected:
+    UniqueGenerator<bsoncxx::types::b_date> _generator;
+};
 
-                millis = (utc_time - epoch).total_milliseconds();
-                break;
-            }
-        }
+class DoubleToIntGenerator : public Generator<int64_t> {
+public:
+    explicit DoubleToIntGenerator(UniqueGenerator<double>& generator) {
+        std::swap(_generator, generator);
+    }
+    int64_t evaluate() override {
+        auto millis = (long long)_generator->evaluate();
+        // BOOST_LOG_TRIVIAL(warning) << " DoubleToIntGenerator " << millis;
+        return millis;
+    }
 
-        if (utc_time == boost::posix_time::not_a_date_time) {
-            try {
-                // Last gasp try to interpret unsigned long long.
-                millis = std::stoull(datetime);
-            } catch (const std::invalid_argument& _) {
-                auto msg = "^RandomDate: Invalid Dateformat '" + datetime + "'";
-                BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax{msg});
-            }
-        }
+protected:
+    UniqueGenerator<double> _generator;
+};
 
-    } else {
-        millis = (default_time - epoch).total_milliseconds();
+class DateStringParserGenerator : public Generator<int64_t> {
+public:
+    explicit DateStringParserGenerator(UniqueGenerator<std::string>& generator) {
+        std::swap(_generator, generator);
     }
-    return std::chrono::milliseconds{millis};
-}
+    int64_t evaluate() override {
+        auto datetime = _generator->evaluate();
+        auto millis = parseStringToMillis(datetime);
+        // BOOST_LOG_TRIVIAL(warning) << " DateStringParserGenerator " << millis;
+        return millis;
+    }
+
+protected:
+    UniqueGenerator<std::string> _generator;
+};
 
 /** `{^RandomDate: {min: "2015-01-01", max: "2015-01-01T23:59:59.999Z"}}` */
 class RandomDateGenerator : public Generator<bsoncxx::types::b_date> {
 public:
     RandomDateGenerator(const Node& node, GeneratorArgs generatorArgs)
         : _rng{generatorArgs.rng},
-          _min{parseDateAsMillis(node["min"].maybe<std::string>().value_or(""))},
-          _max{parseDateAsMillis(node["max"].maybe<std::string>().value_or(""), max_date)} {
-
-        if (_max <= _min) {
-            auto max = node["max"].maybe<std::string>().value_or("");
-            auto min = node["min"].maybe<std::string>().value_or("");
-
-            std::string msg = "^RandomDate: max('" + max + " / " + std::to_string(_max.count()) +
-                "') must be greater than min('" + min + " / " + std::to_string(_min.count()) + "')";
-            BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax{msg});
-        }
-    }
+          _node{node},
+          _minGen{dateGenerator(node["min"], generatorArgs)},
+          _maxGen{dateGenerator(node["max"], generatorArgs, max_date)} {}
 
     bsoncxx::types::b_date evaluate() override {
-        boost::random::uniform_int_distribution<long long> dist{_min.count(), _max.count() - 1};
+        auto min = _minGen->evaluate();
```

Comment:
The error detection happens in evaluate so the test runner needed to be changed to take this into account.